### ISSUE TITLE
Bundle Pristine with adjusted gemspec

### ISSUE THREAD

**Issue by iain** [bundler/bundler#5999]
# Error Report

Summary:

After a bundle pristine with the gem "mail" installed via git, bundle install doesn't run anymore.

Besides the main stacktrace down below, this line is printed somewhere in the middle:

```
Invalid gemspec in [/Users/iain/.gem/ruby/2.4.0/bundler/gems/mail-f0dbe4853856/mail.gemspec]: cannot load such file -- ./lib/mail/version
```

## Questions

Please fill out answers to these questions, it'll help us figure out
why things are going wrong.

- **What did you do?**

  1. Install the mail gem with git as source (see the minimal Gemfile below)
  2. Run `bundle install` to install the gem from git the first time
  3. Run `bundle pristine`
  4. Run `bundle install` (raises the error)

- **What did you expect to happen?**

I didn't change the gem. When I first encountered the issue I had manually changed a different gem, which turned out to be unrelated.

- **What happened instead?**

So it looks like the gem directory is a clone of the git repo, but with an adjusted gemspec. My theory is that bundle pristine does a clean checkout but doesn't re-adjust the gemspec to work again.

I tried to reproduce it with some other gems, but it doesn't occur there. It's probably because the gemspec isn't always different from the version in git.

Workaround A: temporarily remove the gem from the gemfile, running `bundle install` to update the gemfile, then bringing it back, so it reinstalls that version again.

Workaround B: removing the gem manually and have `bundle install` install it again.

- **Have you tried any solutions posted on similar issues in our issue tracker, stack overflow, or google?**

Yeah, found similar sounding bugs, but looking at the stacktrace and the steps, they didn't look to be the same. For instance #5982 has same error description, but I don't think they are related.

- **Have you read our issues document, https://github.com/bundler/bundler/blob/master/doc/contributing/ISSUES.md?**

yes

## Backtrace

```
NoMethodError: undefined method `name' for nil:NilClass
  /Users/iain/.rubies/ruby-2.4.0/lib/ruby/site_ruby/2.4.0/rubygems/stub_specification.rb:158:in `name'
  /Users/iain/.gem/ruby/2.4.0/gems/bundler-1.15.4/lib/bundler/stub_specification.rb:8:in `from_stub'
  /Users/iain/.gem/ruby/2.4.0/gems/bundler-1.15.4/lib/bundler/source/git.rb:319:in `load_gemspec'
  /Users/iain/.gem/ruby/2.4.0/gems/bundler-1.15.4/lib/bundler/source/path.rb:167:in `block in load_spec_files'
  /Users/iain/.gem/ruby/2.4.0/gems/bundler-1.15.4/lib/bundler/source/path.rb:166:in `each'
  /Users/iain/.gem/ruby/2.4.0/gems/bundler-1.15.4/lib/bundler/source/path.rb:166:in `load_spec_files'
  /Users/iain/.gem/ruby/2.4.0/gems/bundler-1.15.4/lib/bundler/source/git.rb:201:in `load_spec_files'
  /Users/iain/.gem/ruby/2.4.0/gems/bundler-1.15.4/lib/bundler/source/path.rb:97:in `local_specs'
  /Users/iain/.gem/ruby/2.4.0/gems/bundler-1.15.4/lib/bundler/source/git.rb:166:in `specs'
  /Users/iain/.gem/ruby/2.4.0/gems/bundler-1.15.4/lib/bundler/lazy_specification.rb:75:in `__materialize__'
  /Users/iain/.gem/ruby/2.4.0/gems/bundler-1.15.4/lib/bundler/spec_set.rb:84:in `block in materialize'
  /Users/iain/.gem/ruby/2.4.0/gems/bundler-1.15.4/lib/bundler/spec_set.rb:81:in `map!'
  /Users/iain/.gem/ruby/2.4.0/gems/bundler-1.15.4/lib/bundler/spec_set.rb:81:in `materialize'
  /Users/iain/.gem/ruby/2.4.0/gems/bundler-1.15.4/lib/bundler/definition.rb:199:in `missing_dependencies'
  /Users/iain/.gem/ruby/2.4.0/gems/bundler-1.15.4/lib/bundler/installer.rb:219:in `block in resolve_if_need'
  /Users/iain/.gem/ruby/2.4.0/gems/bundler-1.15.4/lib/bundler/ui/shell.rb:127:in `with_level'
  /Users/iain/.gem/ruby/2.4.0/gems/bundler-1.15.4/lib/bundler/ui/shell.rb:86:in `silence'
  /Users/iain/.gem/ruby/2.4.0/gems/bundler-1.15.4/lib/bundler/installer.rb:216:in `resolve_if_need'
  /Users/iain/.gem/ruby/2.4.0/gems/bundler-1.15.4/lib/bundler/installer.rb:78:in `run'
  /Users/iain/.gem/ruby/2.4.0/gems/bundler-1.15.4/lib/bundler/installer.rb:24:in `install'
  /Users/iain/.gem/ruby/2.4.0/gems/bundler-1.15.4/lib/bundler/cli/install.rb:68:in `run'
  /Users/iain/.gem/ruby/2.4.0/gems/bundler-1.15.4/lib/bundler/cli.rb:188:in `block in install'
  /Users/iain/.gem/ruby/2.4.0/gems/bundler-1.15.4/lib/bundler/settings.rb:92:in `temporary'
  /Users/iain/.gem/ruby/2.4.0/gems/bundler-1.15.4/lib/bundler/cli.rb:187:in `install'
  /Users/iain/.gem/ruby/2.4.0/gems/bundler-1.15.4/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
  /Users/iain/.gem/ruby/2.4.0/gems/bundler-1.15.4/lib/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'
  /Users/iain/.gem/ruby/2.4.0/gems/bundler-1.15.4/lib/bundler/vendor/thor/lib/thor.rb:387:in `dispatch'
  /Users/iain/.gem/ruby/2.4.0/gems/bundler-1.15.4/lib/bundler/cli.rb:22:in `dispatch'
  /Users/iain/.gem/ruby/2.4.0/gems/bundler-1.15.4/lib/bundler/vendor/thor/lib/thor/base.rb:466:in `start'
  /Users/iain/.gem/ruby/2.4.0/gems/bundler-1.15.4/lib/bundler/cli.rb:13:in `start'
  /Users/iain/.gem/ruby/2.4.0/gems/bundler-1.15.4/exe/bundle:30:in `block in <top (required)>'
  /Users/iain/.gem/ruby/2.4.0/gems/bundler-1.15.4/lib/bundler/friendly_errors.rb:121:in `with_friendly_errors'
  /Users/iain/.gem/ruby/2.4.0/gems/bundler-1.15.4/exe/bundle:22:in `<top (required)>'
  /Users/iain/.gem/ruby/2.4.0/bin/bundle:23:in `load'
  /Users/iain/.gem/ruby/2.4.0/bin/bundle:23:in `<main>'
```

## Environment

```
Bundler   1.15.4
Rubygems  2.6.13
Ruby      2.4.0p0 (2016-12-24 revision 57164) [x86_64-darwin16]
GEM_HOME  /Users/iain/.gem/ruby/2.4.0
GEM_PATH  /Users/iain/.gem/ruby/2.4.0:/Users/iain/.rubies/ruby-2.4.0/lib/ruby/gems/2.4.0
Git       2.14.1
Platform  x86_64-darwin-16
OpenSSL   OpenSSL 1.0.2k  26 Jan 2017
```

## Bundler settings

```
gem.test
  Set for the current user (/Users/iain/.bundle/config): "rspec"
gem.mit
  Set for the current user (/Users/iain/.bundle/config): true
gem.coc
  Set for the current user (/Users/iain/.bundle/config): true
jobs
  Set for your local app (/Users/iain/Teachable/pristien/.bundle/config): 4
```

### Gemfile

```ruby
source "https://rubygems.org"

gem "mail", git: "https://github.com/mikel/mail.git"
```

### Gemfile.lock

```
GIT
  remote: https://github.com/mikel/mail.git
  revision: f0dbe48538560c704e45758f74ea696758aede3a
  specs:
    mail (2.7.0.rc1)
      mini_mime (>= 0.1.1)

GEM
  remote: https://rubygems.org/
  specs:
    mini_mime (0.1.4)

PLATFORMS
  ruby

DEPENDENCIES
  mail!

BUNDLED WITH
   1.15.4
```

PS. The error template is printed twice, do you want me to open another issue for it?

**Comment by segiddins** [comment]
This should be fixed on master! It'll be released in 1.16 very shortly

**Comment by segiddins** [comment]
Can you check with the 1.16 pre-releases? Thanks!
